### PR TITLE
Fixes #4509 Nested imports not resolving correctly

### LINKED ISSUE #4509: AstTypeStubPaths_* test failures
```
AstTypeStubPaths_MergeStubsPath
AstTypeStubPaths_NoStubs
AstTypeStubPaths_MergeStubsPath
AstTypeStubPaths_ExclusiveStubs
```

They all fail like this:
```
System.AggregateException: One or more errors occurred. ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.Fail failed. Failed to get module we just imported
```

### PR DESCRIPTION
**Pull request by zooba**
Fixes #4509 Nested imports not resolving correctly
Corrects assigning submodules to analysis objects
Fixes references to unresolved module names
Fixes comparison of builtin functions

### CODE REVIEW

File: Python/Product/Analysis/Analyzer/DDG.cs [hunk 1]
```diff
@@ -430,30 +430,28 @@ internal class DDG : PythonWalker {
...
+                bool gotAllParents = true;
+                AnalysisValue lastParent = null;
+                remainingParts = name.Split('.');
+                foreach (var part in ModulePath.GetParents(name, includeFullName: false)) {
+                    if (!ProjectState.Modules.TryImport(part, out modRef)) {
+                        gotAllParents = false;
                         break;
                     }
+                    moduleRef = modRef;
+                    (lastParent as BuiltinModule)?.AddChildModule(remainingParts[0], moduleRef.AnalysisModule);
```
Comment (zooba) [inline Python/Product/Analysis/Analyzer/DDG.cs]: This (and the call below) are the critical part of the fix. We may have been resolving `import package.module` into two separate modules (via type stubs) where `package` didn't know about its `.module` member. So now we explicitly associate it when importing.

**Comment by zooba** [comment]
@huguesv Fixes a regression in 15.8 - I'd like to backport it.

File: Python/Product/Analysis/Analyzer/DDG.cs [hunk 2]
```diff
@@ -430,30 +430,28 @@ internal class DDG : PythonWalker {
...
-                            remainingParts = name.Substring(moduleRef.Name.Length + 1).Split('.');
-                        }
-                        _unit.DeclaringModule.AddModuleReference(moduleRef);
-                        return true;
-                    } else {
+                ModuleReference modRef;
+
+                bool gotAllParents = true;
+                AnalysisValue lastParent = null;
+                remainingParts = name.Split('.');
```
Comment (MikhailArkhipov) [inline Python/Product/Analysis/Analyzer/DDG.cs]: What happens if there no dots?

Comment (zooba) [reply Python/Product/Analysis/Analyzer/DDG.cs]: We skip the `foreach` completely and just import it directly.

File: Python/Product/Analysis/Values/BuiltinModule.cs [hunk 3]
```diff
@@ -33,9 +36,23 @@ public BuiltinModule(IPythonModule module, PythonAnalyzer projectState)
 
         public IPythonModule InterpreterModule => _interpreterModule;
 
+        public void AddChildModule(string memberName, AnalysisValue module) {
+            if (_childModules == null) {
+                _childModules = new Dictionary<string, IAnalysisSet>();
+            }
+            if (_childModules.TryGetValue(memberName, out var existing)) {
```
Comment (MikhailArkhipov) [inline Python/Product/Analysis/Values/BuiltinModule.cs]: Sequence is getting popular... Might we worth writing extension to dictionary like `Add<T>(T value, Func<T, T> newValueConstructor);` where the new value constructor is called to make the actual value when existing is there.

**Pull request merged**